libraries/c_sdk/standard/https/src/iot_https_client.c

```diff
@@ -2324,36 +2189,48 @@ static void _sendHttpsRequest( IotTaskPool_t pTaskPool,
 
 /*-----------------------------------------------------------*/
 
+static void _dispatchTaskRoutine( void * pParameters )
+{
+    ( void ) pParameters;
+
+    _httpsRequest_t * pHttpsRequest = NULL;
+
+    for( ; ; )
+    {
+        /* If there are no requests in the dispatch queue, try again. */
+        if( xQueueReceive( dispatchQueue,
+                           &pHttpsRequest,
+                           IOT_HTTPS_QUEUE_RECV_TICKS ) == pdFALSE )
+        {
+            IotLogDebug( ( "No requests to send. Trying again." ) );
+            continue;
+        }
+
+        _sendHttpsRequest( pHttpsRequest );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
 IotHttpsReturnCode_t _scheduleHttpsRequestSend( _httpsRequest_t * pHttpsRequest )
 {
     HTTPS_FUNCTION_ENTRY( IOT_HTTPS_OK );
 
-    IotTaskPoolError_t taskPoolStatus = IOT_TASKPOOL_SUCCESS;
+    BaseType_t queueStatus = pdTRUE;
     _httpsConnection_t * pHttpsConnection = pHttpsRequest->pHttpsConnection;
 
     /* Set the request to scheduled even if scheduling fails. */
     pHttpsRequest->scheduled = true;
 
-    taskPoolStatus = IotTaskPool_CreateJob( _sendHttpsRequest,
-                                            ( void * ) ( pHttpsRequest ),
-                                            &( pHttpsConnection->taskPoolJobStorage ),
-                                            &( pHttpsConnection->taskPoolJob ) );
+    queueStatus = xQueueSendToBack( dispatchQueue, &pHttpsRequest, IOT_HTTPS_QUEUE_SEND_TICKS );
 
-    /* Creating a task pool job should never fail when parameters are valid. */
-    if( taskPoolStatus != IOT_TASKPOOL_SUCCESS )
+    /* Occurs when the queue remains full for #IOT_HTTPS_QUEUE_SEND_TICKS. */
+    if( queueStatus != pdTRUE )
     {
-        IotLogError( "Error creating a taskpool job for request servicing. Error code: %d", taskPoolStatus );
+        IotLogError( "Failed to add request to full dispatch queue. Error code: %d", ( int ) queueStatus );
         HTTPS_SET_AND_GOTO_CLEANUP( IOT_HTTPS_INTERNAL_ERROR );
     }
 
-    taskPoolStatus = IotTaskPool_Schedule( IOT_SYSTEM_TASKPOOL, pHttpsConnection->taskPoolJob, 0 );
```

Comment:
The taskpool scheduling function never had to block because each job itself was a linked list node. I am a little bit concerned about the behavior change here.

Reply:
Completely agree. The reason why I went with a FreeRTOS queue here rather than the original linear container was for the efficiencies it provides:

> When a task attempts to read from an empty queue the task will be placed into the Blocked state (so it is not consuming any CPU time and other tasks can run) until either data becomes available on the queue, or the block time expires.
> 
> When a task attempts to write to a full queue the task will be placed into the Blocked state (so it is not consuming any CPU time and other tasks can run) until either space becomes available in the queue, or the block time expires.
> 
> If more than one task block on the same queue then the task with the highest priority will be the task that is unblocked first.

To emulate the original behavior, it would be best to set the timeout to 0 and return `IOT_HTTPS_INTERNAL_ERROR` when the queue is full. The original task pool returned the same when it ran out of memory to create a task. Slightly differing in behavior, however, I would argue this implementation is better because we can log queue size rather than having the customer estimate how many tasks from the pool were allocated for HTTP before running out of memory.